### PR TITLE
[Improvements] Make melt safer & remove unnecessary array copies

### PR DESCRIPTION
**Pull request by huntabyte**

### CODE REVIEW

File: src/docs/components/preview.svelte
Comment (huntabyte) [inline src/docs/components/preview.svelte:1]: This was also causing a warning to display in the console :)

File: src/docs/components/preview.svelte [hunk 1]
```diff
@@ -85,7 +85,7 @@
 
 	$: files = codingStyleObj !== null ? Object.keys(codingStyleObj) : [];
 
-	let viewCode = false;
```
Comment (huntabyte) [inline src/docs/components/preview.svelte:88]: This was also causing a warning to display in the console :)

File: src/lib/builders/accordion/create.ts [hunk 2]
```diff
@@ -128,11 +130,19 @@ export const createAccordion = <T extends AccordionType = 'single'>(
...
-					const items = Array.from(rootEl.querySelectorAll<HTMLElement>(selector('trigger')));
-					const candidateItems = items.filter((item) => item.dataset.disabled !== 'true');
+					if (!isHTMLElement(rootEl)) return;
+
+					const items = Array.from(rootEl.querySelectorAll(selector('trigger')));
+					const candidateItems = items.filter((item): item is HTMLElement => {
+						if (isHTMLElement(item)) {
+							return item.dataset.disabled !== 'true';
+						}
+						return false;
```
Comment (TGlide) [inline src/lib/builders/accordion/create.ts:144]:
```suggestion
						if (!isHTMLElement(item)) return false;
						return item.dataset.disabled !== 'true';
```

File: src/lib/builders/combobox/create.ts [hunk 3]
```diff
@@ -215,13 +217,16 @@ export function createCombobox<T>(props: CreateComboboxProps<T>) {
 
 							const enabledItems = Array.from(
 								menuEl.querySelectorAll(`${selector('item')}:not([data-disabled])`)
-							) as HTMLElement[];
+							);
```
Comment (TGlide) [inline src/lib/builders/combobox/create.ts:220]: Maybe filter by isHTMLElement here?

File: src/lib/builders/pin-input/create.ts [hunk 4]
```diff
@@ -16,12 +18,14 @@ import type { CreatePinInputProps } from './types';
...
 const { name, selector } = createElHelpers<'input' | 'hidden-input'>('pin-input');
 
-const getInputs = (node: HTMLElement) => {
+const getInputs = (node: HTMLInputElement) => {
 	const rootEl = node.closest(selector());
-	if (!rootEl) return { inputs: null, el: node as HTMLInputElement, elIndex: -1 };
-	const inputs = Array.from(rootEl.querySelectorAll(selector('input'))) as HTMLInputElement[];
+	if (!isHTMLElement(rootEl)) return { inputs: null, el: node, elIndex: -1 };
+	const inputs = Array.from(rootEl.querySelectorAll(selector('input'))).filter(
+		(input): input is HTMLInputElement => isHTMLInputElement(input)
```
Comment (TGlide) [inline src/lib/builders/pin-input/create.ts:25]: Do you need the `input is HTMLInputElement` part? That should already be the return type of `isHTMLInputElement` 🤔

File: src/lib/builders/slider/create.ts [hunk 5]
```diff
@@ -101,10 +102,9 @@ export const createSlider = (props: CreateSliderProps = defaults) => {
 		const root = getElementByMeltId(ids.root);
 		if (!root) return null;
 
-		const thumbs = Array.from(root.querySelectorAll('[data-melt-part="thumb"]')).filter(
-			Boolean
-		) as Array<HTMLElement>;
-		return thumbs;
+		return Array.from(root.querySelectorAll('[data-melt-part="thumb"]')).filter(
+			(thumb): thumb is HTMLElement => isHTMLElement(thumb)
```
Comment (TGlide) [inline src/lib/builders/slider/create.ts:106]: Same here

File: src/lib/builders/slider/create.ts [hunk 6]
```diff
@@ -280,11 +281,13 @@ export const createSlider = (props: CreateSliderProps = defaults) => {
...
 
-			const sliderEl = getElementByMeltId($root['data-melt-id']) as HTMLElement;
+			const sliderEl = getElementByMeltId($root['data-melt-id']);
 			const closestThumb = getClosestThumb(e);
 			if (!closestThumb || !sliderEl) return;
 
-			if (!sliderEl.contains(e.target as HTMLElement)) return;
+			const target = e.target;
+			if (!isHTMLElement(target)) return;
+			if (!sliderEl.contains(target)) return;
```
Comment (TGlide) [inline src/lib/builders/slider/create.ts:290]: These can be in the same if statement

File: src/lib/builders/tabs/create.ts [hunk 7]
```diff
@@ -130,35 +131,44 @@ export function createTabs(props?: CreateTabsProps) {
...
+					const rootEl = el.closest(selector());
 
-					if (!rootEl || !tabValue) return;
+					if (!isHTMLElement(rootEl) || !tabValue) return;
 
 					const $options = get(options);
 
-					const triggers = Array.from(rootEl.querySelectorAll('[role="tab"]')) as HTMLElement[];
+					const triggers = Array.from(rootEl.querySelectorAll('[role="tab"]'));
 					const enabledTriggers = triggers.filter((el) => !el.hasAttribute('data-disabled'));
```
Comment (TGlide) [inline src/lib/builders/tabs/create.ts:143]: We could possibly filter by htmlElement here and remove the inner if statements. Same for other cases where this happens.

File: src/lib/builders/toggle-group/create.ts [hunk 8]
```diff
@@ -118,11 +122,11 @@ export function createToggleGroup(props: CreateToggleGroupProps = {}) {
 					const el = e.currentTarget;
 					if (!isHTMLElement(el)) return;
 
-					const root = el.closest<HTMLElement>(selector());
+					const root = el.closest(selector());
 					if (!isHTMLElement(root)) return;
 
 					const items = Array.from(
-						root.querySelectorAll<HTMLElement>(selector('item') + ':not([data-disabled])')
+						root.querySelectorAll(selector('item') + ':not([data-disabled])')
```
Comment (TGlide) [inline src/lib/builders/toggle-group/create.ts:129]: Filter by HTMLElement

File: src/lib/internal/helpers/event.ts [hunk 9]
```diff
@@ -47,7 +47,7 @@ export function addEventListener(
 	handler: EventListenerOrEventListenerObject,
 	options?: boolean | AddEventListenerOptions
 ) {
-	const events = Array.isArray(event) ? [...event] : [event];
+	const events = Array.isArray(event) ? event : [event];
```
Comment (TGlide) [inline src/lib/internal/helpers/event.ts:50]: Scary!

Comment (TGlide) [reply src/lib/internal/helpers/event.ts:50]: (should be fine tho)

**Review by TGlide — CHANGES_REQUESTED** [review]
Awesome! Just some small changes needed 😄

File: src/lib/builders/pin-input/create.ts (hunk 4 above)
Comment (huntabyte) [reply src/lib/builders/pin-input/create.ts:25]: I know filters are a bit weird and I couldn't get it to type properly without it, but let me check again :)

File: src/lib/internal/helpers/event.ts (hunk 9 above)
Comment (huntabyte) [reply src/lib/internal/helpers/event.ts:50]: 🤣

**Pull request merged**